### PR TITLE
fix(linker): prevent pollution of empty embeddedView context

### LINKED ISSUE #10045: TemplateRef shares context between instances when none is provided
**I'm submitting a ...**  (check one with "x")
```
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
`angular/modules/@angular/core/src/linker/template_ref.ts` currently uses a single object for `EMPTY_CONTEXT`. The problem is when `ViewContainerRef #createEmbededView` is called without providing a context object the shared one is used causing pollution between all instances utilizing the context.

**Expected/desired behavior**
Object sharing doesn't happen between instances relying on the context.

Maybe some solutions could be:
* `TemplateRef_<C>` instantiates a new object when none is provided
* `ViewContainerRef #createEmbededView` makes context mandatory and `TemplateRef_<C>` stops checking isBlank of context.

**Reproduction of the problem**
If the current behavior is a bug or you can illustrate your feature request better with an example, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5).

https://plnkr.co/edit/UhGsjLaHxiwI0n0nTYtN?p=preview
//TypeError: Cannot read property 'name' of undefined on second instance

Simplified Plunk: http://plnkr.co/edit/CQQembTFT0nOxGwCzSgc?p=preview
You can see the sharing of the context, `test2` is listed twice in first block

**What is the expected behavior?**
No sharing of the context between instances

**What is the motivation / use case for changing the behavior?**
Errors

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.4
 
* **Browser:** [all] 
 
* **Language:** [TypeScript 1.8.10]

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Bugfix
```

**What is the current behavior?** (You can also link to an open issue here)

It is possible to pollute / share context of embeddedView, see #10045

**Does this PR introduce a breaking change?** (check one with "x")

```
[ x] No
```

**Other information**:

This is a super-corner case but since someone bumped into it...
Fixes #10045

### CODE REVIEW

File: modules/@angular/core/src/linker/template_ref.ts [hunk 1]
```diff
@@ -52,7 +50,7 @@ export class TemplateRef_<C> extends TemplateRef<C> {
     var view: AppView<C> = this._viewFactory(
         this._appElement.parentView.viewUtils, this._appElement.parentInjector, this._appElement);
     if (isBlank(context)) {
-      context = <any>EMPTY_CONTEXT;
+      context = <any>new Object();
     }
     view.create(context, null, null);
```
Comment (vicb) [inline modules/@angular/core/src/linker/template_ref.ts]: `context || {}`

File: modules/@angular/core/test/linker/integration_spec.ts [hunk 2]
```diff
@@ -470,6 +470,27 @@ function declareTests({useJit}: {useJit: boolean}) {
                    });
              }));
 
+      it('should not share empty context for template directives',
```
Comment (vicb) [inline modules/@angular/core/test/linker/integration_spec.ts]: could you please add a comment with the link to the GH issue ?

File: modules/@angular/core/src/linker/template_ref.ts (hunk 1 above)
Comment (pkozlowski-opensource) [inline modules/@angular/core/src/linker/template_ref.ts]: Done

File: modules/@angular/core/test/linker/integration_spec.ts (hunk 2 above)
Comment (pkozlowski-opensource) [inline modules/@angular/core/test/linker/integration_spec.ts]: Done

**Pull request merged**